hal/armv7m/mpu.c

```diff
@@ -0,0 +1,333 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Phoenix-RTOS loader
+ *
+ * MPU routines
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <lib/lib.h>
+#include <lib/log.h>
+#include <lib/errno.h>
+#include <hal/hal.h>
+#include <syspage.h>
+#include "mpu.h"
+
+enum {
+	mpu_fError = 0x01,
+	mpu_fConfigured = 0x04,
+	mpu_fValid = 0x08,
+};
+
+
+struct {
+	syspage_hal_t hal;
+	u8 mapsCnt;
+	u8 regCnt;
+	u8 regMax;
+	u8 flags;
+} mpu_common;
+
+
+/* clang-format off */
+
+/* Binary count trailing zeros */
+__attribute__((always_inline)) static inline u8 _ctz(u32 val)
+{
+	asm volatile(" \
+		rbit %0, %0; \
+		clz %0, %0"
+		: "+l"(val));
+	return val;
+}
+
+/* clang-format on */
+
+
+/* Get MPU type and features */
+static int mpu_type(void)
+{
+	return *(volatile u32 *)0xe000ed90;
```

Comment:
Use `*(base + register)` to access registers. See kernel/device drivers for examples